src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PagePropertyBinderFactory.cs

```diff
@@ -0,0 +1,188 @@
+                return BindPropertiesAsync(parameterBinder, pageContext, instance, propertiesToBind);
+            };
+        }
+
+        private static async Task BindPropertiesAsync(
+            ParameterBinder parameterBinder,
+            PageContext pageContext,
+            object instance,
+            IList<PropertyBindingInfo> propertiesToBind)
+        {
+            var valueProvider = await GetCompositeValueProvider(pageContext);
+            for (var i = 0; i < propertiesToBind.Count; i++)
+            {
+                var propertyBindingInfo = propertiesToBind[i];
+                var modelBindingResult = await parameterBinder.BindModelAsync(
+                    pageContext, 
+                    valueProvider, 
+                    propertyBindingInfo.ParameterDescriptor);
+                if (modelBindingResult.IsModelSet)
+                {
+                    var modelMetadata = propertyBindingInfo.ModelMetadata;
+                    PropertyValueSetter.SetValue(
+                        modelMetadata,
+                        instance,
+                        modelBindingResult.Model);
+                }
+            }
+        }
+
+        private static IList<PropertyBindingInfo> GetPropertiesToBind(
+            IModelMetadataProvider modelMetadataProvider,
+            TypeInfo handlerSourceTypeInfo)
+        {
+            var handlerType = handlerSourceTypeInfo.AsType();
+            var handlerMetadata = modelMetadataProvider.GetMetadataForType(handlerType);
+            if (handlerMetadata.Properties.Count == 0)
+            {
+                return EmptyArray<PropertyBindingInfo>.Instance;
+            }
+
+            PropertyHelper[] propertyInfo = null;
+
+            var propertyBindingInfo = new List<PropertyBindingInfo>();
+            for (var i = 0; i < handlerMetadata.Properties.Count; i++)
+            {
+                var propertyMetadata = handlerMetadata.Properties[i];
+                if (propertyMetadata.IsReadOnly)
+                {
+                    continue;
+                }
+
+                BindingInfo bindingInfo;
+                if (propertyMetadata.BindingSource != null)
+                {
+                    bindingInfo = new BindingInfo()
+                    {
+                        BinderModelName = propertyMetadata.BinderModelName,
+                        BinderType = propertyMetadata.BinderType,
+                        BindingSource = propertyMetadata.BindingSource,
+                        PropertyFilterProvider = propertyMetadata.PropertyFilterProvider,
+                    };
+                }
+                else
+                {
+                    if (propertyInfo == null)
+                    {
+                        propertyInfo = PropertyHelper.GetVisibleProperties(type: handlerSourceTypeInfo.AsType());
+                    }
+                    bindingInfo = GetBindingInfoFromProperty(propertyInfo, propertyMetadata);
```

Comment:
Properties with vanilla `ModelBinderAttribute` don't have a `BindingSource` set. Is there another way to infer this?